packages/debug/src/browser/debug-session-manager.ts

```diff
@@ -278,51 +278,69 @@ export class DebugSessionManager {
             const restart = event.body && event.body.restart;
             if (restart) {
                 // postDebugTask isn't run in case of auto restart as well as preLaunchTask
-                this.doRestart(session, restart);
+                this.doRestart(session, !!restart);
             } else {
-                session.terminate();
+                await session.disconnect(false, () => this.debug.terminateDebugSession(session.id));
                 await this.runTask(session.options.workspaceFolderUri, session.configuration.postDebugTask);
             }
         });
-        session.on('exited', () => this.destroy(session.id));
-        session.start().then(() => this.onDidStartDebugSessionEmitter.fire(session));
+
+        session.on('exited', async event => {
+            await session.disconnect(false, () => this.debug.terminateDebugSession(session.id));
+        });
+
+        session.onDispose(() => this.cleanup(session));
+        session.start().then(() => this.onDidStartDebugSessionEmitter.fire(session)).catch(e => session.stop(false, () => this.debug.terminateDebugSession(session.id)));
```

Comment:
Catching the start error and trying to terminate it afterwards will lead to unexpected behavior. `session.stop` awaits a `sendRequest` call just before executing the callback. When the session errors on startup, the `sendRequest` call will never complete, and we are left with a dangling debug session. You test this by installing the current version of `cortex-debug` from OVSX and running this debug config:

```json
{
  "cwd": "${workspaceRoot}",
  "name": "Cortex-Debug",
  "request": "launch",
  "type": "cortex-debug",
  "executable": "",
  "servertype": "openocd",
  "serverpath": "",
  "armToolchainPath": "",
  "configFiles": [
    ""
  ]
}
```

`session.stop` Is called and will stop it's execution at the following line, which never completes:
https://github.com/eclipse-theia/theia/blob/bbba07d45ffa63b7c4edb767f66f2487d6b7d0d2/packages/debug/src/browser/debug-session.tsx#L341

Reply:
I would suspect the problem is that we don't know the state of the connection to the debug adapter if the session start fails. My approach would be to make sending messages fail when the connection to the debugger is closed.

Reply:
Hmh...debugging through the scenario with cortex-debug, the creation of the session does _not_ fail. However, sending the "disconnect" request never resolves nor rejects.

Reply:
Here's the sequenced I get when runng the debug config: 

```
06b9e88e theia -> adapter: {"seq":1,"type":"request","command":"initialize","arguments":{"clientID":"Theia","clientName":"Theia IDE","adapterID":"cortex-debug","locale":"en-US","linesStartAt1":true,"columnsStartAt1":true,"pathFormat":"path","supportsVariableType":false,"supportsVariablePaging":false,"supportsRunInTerminalRequest":true}}
06b9e88e theia -> adapter: {"seq":2,"type":"request","command":"get-arguments"}
06b9e88e theia <- adapter: {"seq":1,"type":"response","request_seq":1,"command":"initialize","success":true,"body":{"supportsHitConditionalBreakpoints":true,"supportsConfigurationDoneRequest":true,"supportsConditionalBreakpoints":true,"supportsFunctionBreakpoints":true,"supportsEvaluateForHovers":true,"supportsSetVariable":true,"supportsRestartRequest":true,"supportsGotoTargetsRequest":true,"supportSuspendDebuggee":true,"supportTerminateDebuggee":true}}
06b9e88e theia -> adapter: {"seq":3,"type":"request","command":"launch","arguments":{"cwd":"c:\\Users\\thomas\\code\\chedev\\theia","name":"Cortex-Debug","request":"launch","type":"cortex-debug","executable":"","servertype":"openocd","serverpath":null,"armToolchainPath":"","configFiles":[""],"noDebug":false,"gdbServerConsolePort":55878,"debuggerArgs":[],"demangle":true,"swoConfig":{"enabled":false,"decoders":[],"cpuFrequency":0,"swoFrequency":0,"source":"probe"},"rttConfig":{"enabled":false,"decoders":[]},"graphConfig":[],"preLaunchCommands":[],"postLaunchCommands":[],"preAttachCommands":[],"postAttachCommands":[],"preRestartCommands":[],"postRestartCommands":[],"searchDir":[],"toolchainPath":null,"toolchainPrefix":"arm-none-eabi","gdbPath":null,"extensionPath":"c:/Users/thomas/.theia/extensions/marus25.cortex-debug-0.4.7/extension","flattenAnonymous":false,"registerUseNaturalFormat":true,"variableUseNaturalFormat":true}}
06b9e88e theia <- adapter: {"seq":2,"type":"response","request_seq":2,"command":"get-arguments","success":false,"message":"TypeError: Cannot read property 'customRequest' of undefined\n    at q.customRequest (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:31666)\n    at q.dispatchRequest (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:1:18100)\n    at q.handleMessage (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:61573)\n    at q._handleData (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:63297)\n    at Socket.<anonymous> (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:61907)\n    at Socket.emit (events.js:315:20)\n    at addChunk (_stream_readable.js:295:12)\n    at readableAddChunk (_stream_readable.js:271:9)\n    at Socket.Readable.push (_stream_readable.js:212:10)\n    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)","body":{"error":{"id":1104,"format":"{_stack}","variables":{"_exception":"Cannot read property 'customRequest' of undefined","_stack":"TypeError: Cannot read property 'customRequest' of undefined\n    at q.customRequest (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:31666)\n    at q.dispatchRequest (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:1:18100)\n    at q.handleMessage (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:61573)\n    at q._handleData (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:63297)\n    at Socket.<anonymous> (c:\\Users\\thomas\\.theia\\extensions\\marus25.cortex-debug-0.4.7\\extension\\dist\\debugadapter.js:9:61907)\n    at Socket.emit (events.js:315:20)\n    at addChunk (_stream_readable.js:295:12)\n    at readableAddChunk (_stream_readable.js:271:9)\n    at Socket.Readable.push (_stream_readable.js:212:10)\n    at Pipe.onStreamRead (internal/stream_base_commons.js:186:23)"},"sendTelemetry":true}}}
06b9e88e theia <- adapter: {"seq":3,"type":"response","request_seq":3,"command":"launch","success":false,"message":"Unable to find executable file at .","body":{"error":{"id":103,"format":"Unable to find executable file at .","showUser":true}}}
06b9e88e theia -> adapter: {"seq":4,"type":"request","command":"disconnect","arguments":{"restart":false}}
```

To me, this looks like a misbehaviour of the debug adapter: it remains running, but does not send an answer to the disconnect request. I noticed that we're not reacting to the failed launch request, but even if we detect that, we'd still have to disconnect from the debug adapter somehow, so the problem would be the same.

Reply:
Hm, interesting. I wonder how vscode is dealing with this. It fails with the same error, but seems to resolve the issue (at least in the frontend) gracefully. I don't know what actually happens in the backend though, and whether everything is terminating there correctly.

Reply:
I've seen this before,  vscode has a lot more guards against misbehaving adapters

Reply:
I've introduced a timeout when sending the "disconnect" event. That should shield us from the debugger not responding to the event.